### PR TITLE
feat(sdk): add local execution output collection #localexecution

### PR DESCRIPTION
**Pull request by connor-mccarthy**
**Description of your changes:**
Adds the code that collects and surfaces the outputs from locally executed tasks on the `task.output` and `task.outputs['<key>']`.

This constitutes E2E local execution for the subprocess task handler. For this reason, we also now include E2E tests.

Forthcoming PRs will add the container task handler, logging, and a few more changes.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/dsl/types/artifact_types.py [hunk 1]
```diff
@@ -89,7 +89,8 @@ def _get_path(self) -> Optional[str]:
             return _MINIO_LOCAL_MOUNT_PREFIX + self.uri[len('minio://'):]
         elif self.uri.startswith('s3://'):
             return _S3_LOCAL_MOUNT_PREFIX + self.uri[len('s3://'):]
-        return None
+        # uri == path for local execution
+        return self.uri
```
Comment (connor-mccarthy) [inline sdk/python/kfp/dsl/types/artifact_types.py:93]: @chensun, WDYT about this change?

It is a behavior change. If a user depends on `.path` being `None`, their code's behavior may change. This doesn't affect code paths that write to `.path` (since you cannot write to `None`), but it would affect code like this:

```python
@dsl.component
def my_comp(a: Artifact):
    if a.path is None: # prev evaluates to True, now would evaluate to False
        persist('some artifact contents', a.uri)
    else:
        persist('some artifact contents', a.path)
```

The alternative is to set some sort of special flag on both input and output artifacts to allow `.path` to return the "local uri" when executing locally:

```python
# implementation code
for k, maybe_artifact in user_kwargs.items():
    if running_locally and isinstance(maybe_artifact, dsl.Artifact):
        maybe_artifact._use_uri_as_path = True

# then change `.path` to respect the `._use_uri_as_path` flag
```

WDYT? The reason for not immediately proceeding with the alternative implementation above is that:
- The current implementation's behavior may be acceptable in a minor version bump with release notes.
- The solution above is somewhat hacky and it's challenging to exhaustively reason about whether it will play nicely with user local execution code or, eventually, input artifact constants.

I can go either way on this.

File: sdk/python/kfp/local/executor_output_utils.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        executor_output=executor_output,
+        component_spec=component_spec,
+    )
+
+    # collect parameter outputs from executor output
+    output_parameters = {
+        param_name: pb2_value_to_python(value)
+        for param_name, value in executor_output.parameter_values.items()
+    }
+    # process the special case of protobuf ints
```
Comment (connor-mccarthy) [inline sdk/python/kfp/local/executor_output_utils.py:76]: This is related to [this previous comment](https://github.com/kubeflow/pipelines/pull/10302#discussion_r1428394031). The current implementation in this PR is the correct way to handle this, IMO.

File: sdk/python/kfp/dsl/types/artifact_types.py (hunk 1 above)
Comment (chensun) [reply sdk/python/kfp/dsl/types/artifact_types.py:93]: > If a user depends on .path being None, their code's behavior may change.

This sounds fine to me. I don't perceive `.path` being `None` an existing scenario until very recent changes for local execution.

File: sdk/python/kfp/local/e2e_test.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    def test_str_input(self, runner):
+        local.init(runner=runner)
+
+        @dsl.component
+        def identity(x: str) -> str:
+            return x
+
+        actual = identity(x='hello').output
+        expected = 'hello'
+        self.assertIsInstance(actual, str)
```
Comment (chensun) [inline sdk/python/kfp/local/e2e_test.py:59]: nit: seems a bit redundant to me as the equality test next should cover this as well. 
But I don't have strong opinion, feel free to keep it.

File: sdk/python/kfp/local/executor_output_utils.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    outputs_typed_int = [
+        output_param_name for output_param_name, parameter_spec in
+        component_spec.output_definitions.parameters.items()
+        if parameter_spec.parameter_type ==
+        pipeline_spec_pb2.ParameterType.ParameterTypeEnum.NUMBER_INTEGER
+    ]
+    output_parameters = cast_ints_as_floats_to_ints(
```
Comment (chensun) [inline sdk/python/kfp/local/executor_output_utils.py:83]: nit: naming of `cast_ints_as_floats_to_ints` is a bit hard to follow. one possibly cleaner way is to combine this and the code block above to have some function like `casting_protobuf_numbers`.

**Review by chensun — COMMENTED** [review]
/lgtm

only a few nitpicks, feel free to ignore

File: sdk/python/kfp/local/e2e_test.py (hunk 3 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/local/e2e_test.py:59]: I added [a comment](https://github.com/kubeflow/pipelines/blob/4e3699e4dcbbe3a84204213c79c71d9c66c4d418/sdk/python/kfp/local/e2e_test.py#L33-L38) to the top of the file on why this is needed. Added to the top, since it applies for every assertion.

File: sdk/python/kfp/local/executor_output_utils.py (hunk 4 above)
Comment (connor-mccarthy) [reply sdk/python/kfp/local/executor_output_utils.py]: Updated in 958362c3353224d4bda0ad588ede1f8a406d60b3.

I also noticed that this casting is needed on the input side. [3f86da7](https://github.com/kubeflow/pipelines/pull/10325/commits/3f86da7b1a53e25c6b1c234a2460e8df084fef88) adds casting when going from the ExecutorInput to JSON. This is consistent with the remote BE behavior.

This is needed because of the case like:

```python
@dsl.component
def my_comp(x: int) -> int:
    return x
```

If `x` goes in as a `float` (default representation in protobuf Value), it will be returned as a `float` and throw a runtime exception, since it expects an `int` return. Also related to [previous comment](https://github.com/kubeflow/pipelines/pull/10302#discussion_r1428394031).

**Review by chensun — APPROVED** [review]
/lgtm
/approve

Thanks!

**Comment by connor-mccarthy** [comment]
Merging per [prior LGTM](https://github.com/kubeflow/pipelines/pull/10325#pullrequestreview-1785052799). Changes since the LGTM include rebasing and removing 2 duplicate LOC causing a false positive test failure.

`kfp-kubernetes-execution-tests` test failures are independent of this PR. The pipeline is hanging in a "Running" state, despite inner tasks all succeeding. Investigating separately.
